File: Backend/app/schemas/journal_entry.py
"""Manual Journal Entry schemas — design doc §5.4.

Debit/credit arrive as `Money` (string on the wire, exact `Decimal` once
parsed) so the client can never accidentally send a float that silently
loses precision before it even reaches the engine's balance check.
"""
from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.models.enums import JournalEntrySourceType, JournalEntryStatus
from app.schemas.common import Money


class JournalEntryLineIn(BaseModel):
    account_id: int
    partner_id: int | None = None
    analytic_account_id: int | None = None
    label: str | None = Field(default=None, max_length=255)
    debit: Money = Decimal("0")
    credit: Money = Decimal("0")

    @model_validator(mode="after")
    def _shape(self):
        # Client-side convenience only — the engine's own validators (rules
        # 2/3/4) are the actual gate; this just gives a 422 instead of a 409
        # for the most obvious malformed-input case.
        if self.debit < 0 or self.credit < 0:
            raise ValueError("debit and credit cannot be negative")
        return self


class JournalEntryCreate(BaseModel):
    journal_id: int
    entry_date: date
    reference: str | None = Field(default=None, max_length=128)
    narration: str | None = None
    lines: list[JournalEntryLineIn] = Field(min_length=1)


class JournalEntryUpdate(BaseModel):
    journal_id: int | None = None
    entry_date: date | None = None
    reference: str | None = Field(default=None, max_length=128)
    narration: str | None = None
    lines: list[JournalEntryLineIn] | None = Field(default=None, min_length=1)


class JournalEntryLineOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    line_no: int
    account_id: int
    partner_id: int | None
    analytic_account_id: int | None
    label: str | None
    debit: Money
    credit: Money


class JournalEntryOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    entry_number: str
    journal_id: int
    entry_date: date
    reference: str | None
    narration: str | None
    status: JournalEntryStatus
    source_type: JournalEntrySourceType
    source_document_id: int | None
    source_payment_id: int | None
    total_debit: Money
    total_credit: Money
    posted_at: datetime | None
    posted_by: int | None
    lines: list[JournalEntryLineOut]

File: Frontend/src/types/journalEntry.ts
export type JournalEntryStatus = 'DRAFT' | 'POSTED' | 'CANCELLED'
export type JournalEntrySourceType = 'MANUAL' | 'VENDOR_BILL' | 'CUSTOMER_INVOICE' | 'PAYMENT'

export interface JournalEntryLine {
  id: number
  line_no: number
  account_id: number
  partner_id: number | null
  analytic_account_id: number | null
  label: string | null
  debit: string
  credit: string
}

export interface JournalEntry {
  id: number
  entry_number: string
  journal_id: number
  entry_date: string
  reference: string | null
  narration: string | null
  status: JournalEntryStatus
  source_type: JournalEntrySourceType
  source_document_id: number | null
  source_payment_id: number | null
  total_debit: string
  total_credit: string
  posted_at: string | null
  posted_by: number | null
  lines: JournalEntryLine[]
}

export interface JournalEntryLineInput {
  account_id: number
  partner_id?: number | null
  analytic_account_id?: number | null
  label?: string | null
  debit: string
  credit: string
}

export interface JournalEntryInput {
  journal_id: number
  entry_date: string
  reference?: string | null
  narration?: string | null
  lines: JournalEntryLineInput[]
}
